### PR TITLE
[Consensus] add type and interface for transaction voting

### PR DESCRIPTION
**Pull request by mwtian**
## Description 

Add block variant that supports voting.
Add method to TransactionVerifier trait, for calling into the voting logic.

## Test plan 

CI

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:
- [ ] REST API:

### CODE REVIEW

File: consensus/core/src/block.rs [hunk 1]
```diff
@@ -50,16 +50,30 @@ impl Transaction {
...
 /// equivocate.
+#[allow(private_interfaces)]
 #[derive(Clone, Deserialize, Serialize)]
 #[enum_dispatch(BlockAPI)]
 pub enum Block {
     V1(BlockV1),
+    V2(BlockV2),
 }
 
+#[allow(private_interfaces)]
```
Comment (akichidis) [inline consensus/core/src/block.rs:76]: Out of curiosity, why is this needed now? Isi t because we reduced the visibility of `BlockV1/2` ?

File: consensus/core/src/transaction.rs [hunk 2]
```diff
@@ -229,6 +234,10 @@ impl TransactionVerifier for NoopTransactionVerifier {
     fn verify_batch(&self, _batch: &[&[u8]]) -> Result<(), ValidationError> {
         Ok(())
     }
+
+    fn vote_batch(&self, _batch: &[&[u8]]) -> Vec<TransactionIndex> {
+        vec![]
+    }
```
Comment (akichidis) [inline consensus/core/src/transaction.rs:240]: Do we need to return a here a rejection reason as well? I know that as part of the protocol we don't look to include the rejection reasons in the DAG (although it would be convenient now that we won't have quorum driver gathering the responses) , but it might be useful? I guess we can deal with that if needed later.

**Review by akichidis — APPROVED** [review]

File: crates/sui-core/src/consensus_validator.rs [hunk 3]
```diff
@@ -152,6 +152,10 @@ impl TransactionVerifier for SuiTxValidator {
         self.validate_transactions(txs)
             .map_err(|e| ValidationError::InvalidTransaction(e.to_string()))
     }
+
+    fn vote_batch(&self, _batch: &[&[u8]]) -> Vec<TransactionIndex> {
```
Comment (aschran) [inline crates/sui-core/src/consensus_validator.rs:156]: not sure yet how you're planning to use this but are you sure return type shouldn't be wrapped in a Result?

Comment (aschran) [reply crates/sui-core/src/consensus_validator.rs:156]: ok I now see there's a TODO on this interface elsewhere

**Review by aschran — APPROVED** [review]

File: crates/sui-core/src/consensus_validator.rs (hunk 3 above)
Comment (mwtian) [reply crates/sui-core/src/consensus_validator.rs:156]: Yes this can be useful. Expanded the TODO.

File: consensus/core/src/block.rs [hunk 4]
```diff
@@ -50,16 +50,31 @@ impl Transaction {
...
 }
 
+/// Index of a transaction in a block.
+pub type TransactionIndex = u16;
+
+/// Votes on transactions in a specific block.
+/// Reject votes are explicit. The rest of transactions in the block receive implicit accept votes.
+// TODO: look into making fields `pub`.
+#[derive(Clone, Deserialize, Serialize)]
+pub(crate) struct BlockTransactionVotes {
```
Comment (arun-koshy) [inline consensus/core/src/block.rs:60]: Just `TransactionVotes` seems cleaner and we don't have conflict with other transaction votes outside of this

**Review by arun-koshy — APPROVED** [review]

File: consensus/core/src/block.rs (hunk 4 above)
Comment (mwtian) [reply consensus/core/src/block.rs:60]: I think it is useful to indicate the votes are for a particular Block. Let's see if the prefix will turn out to be redundant.
